### PR TITLE
Add iPAddress support

### LINKED ISSUE #12: Support iPAddress in subjectAltName extensions
It is a thing.  See http://tools.ietf.org/html/rfc5280#section-4.2.1.6 I guess.

### PR DESCRIPTION
**Pull request by hynek**
Fixes #12 

This implements basic support for iPAdress SANs. There is no higher/public level API to use them yet but I think we can/should review this first.

### CODE REVIEW

**Comment by aronbierbaum** [comment]
Is the idea that you could pass an IP address to verify_hostname? Have you started working on that aspect yet?

**Comment by cheungnj** [comment]
I don't think this works if you pass an IP address to `verify_hostname` but I don't know if that's the expected behavior or not. Also I could really use this PR and the ability to pass IP addresses to `verify_hostname`. 😄

**Comment by twm** [comment]
I'd like to add support for `https://<ip-addr>` in Twisted. https://github.com/twisted/twisted/pull/1028 was the first step (permitting such URLs syntactically), but we need the higher-level API which would build on this PR to validate certs.

I've reviewed the changes here, which seem correct. Is there anything I can do to help this along? I could provide a PR which adds the higher-level API once this one has been merged.

**Comment by hynek** [comment]
The problem was that it was sort of blocked by Alex who thinks that we shouldn’t add more feature to s_i and try to use OpenSSL’s features instead.

But since it appears that cryptography/pyOpenSSL will grow verification itself using OpenSSL’s mechanisms, I feel like we should follow through on this anyway.

**Comment by twm** [comment]
Okay, so you want to merge this? Is there anything I can do to help make that happen?

**Comment by hynek** [comment]
Currently there’s the open question raised by @tiran whether we shouldn’t use `inet_pton` instead of `ipaddress`. The big question for me is how compatible that would be.

Also I’ve got quite the FOSS backlog right now. :'(

**Comment by hynek** [comment]
For reference: https://github.com/python/cpython/blob/master/Lib/ssl.py#L240

**Comment by twm** [comment]
The goal of using `inet_pton` instead of `ipaddress` being to avoid adding the dependency on Python 2.7?

Twisted [carries a monkey patch](https://github.com/twisted/twisted/blob/ee535041258e7ef0b3223d2e12cd9aaa0bc2289f/src/twisted/python/compat.py#L198-L207) for those functions relating to IPv6 support, so they were at least historically flaky. [#5895](https://twistedmatrix.com/trac/ticket/5895) aims to deprecate and remove.

The CPython implementation you link notes that it only supports IPv6 addresses on platforms with IPv6 support. I wouldn't expect loading a certificate to fail due to platform support in that way, so that seems like a problem.

**Comment by hynek** [comment]
> The goal of using inet_pton instead of ipaddress being to avoid adding the dependency on Python 2.7?

And speed.

> The CPython implementation you link notes that it only supports IPv6 addresses on platforms with IPv6 support. I wouldn't expect loading a certificate to fail due to platform support in that way, so that seems like a problem.

Are there any relevant platforms left that don’t support IPv6? Even RHEL 4 has IPv6.

Also that is solvable with falling back to inet_aton.

**Comment by twm** [comment]
> Are there any relevant platforms left that don’t support IPv6? Even RHEL 4 has IPv6.

I don't personally go far afield, platform-wise, so I created a [Twisted PR which removes the monkey patch](https://github.com/twisted/twisted/pull/1082) to see what the build fleet things.

The answer appears to be yes. The Windows 7 build failed like so:

```
Traceback (most recent call last):
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\python\usage.py", line 411, in <lambda>
    fn = lambda name, value, m=method: m(value)
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\application\app.py", line 539, in opt_reactor
    installReactor(shortName)
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\application\reactors.py", line 80, in installReactor
    for installer in getReactorTypes():
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\plugin.py", line 213, in getPlugins
    allDropins = getCache(package)
--- <exception caught here> ---
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\plugin.py", line 171, in getCache
    provider = pluginModule.load()
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\python\modules.py", line 392, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\python\reflect.py", line 308, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\python\reflect.py", line 255, in _importAndCheckStack
    reraise(excValue, excTraceback)
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\plugins\twisted_core.py", line 6, in <module>
    from twisted.internet.endpoints import (
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\internet\endpoints.py", line 29, in <module>
    from twisted.internet.abstract import isIPv6Address, isIPAddress
  File "C:\buildslave\windows7-64-py2_7\Twisted\build\py27-alldeps-nocov-windows\lib\site-packages\twisted\internet\abstract.py", line 11, in <module>
    from socket import AF_INET, AF_INET6, inet_pton, error
exceptions.ImportError: cannot import name inet_pton
```

**Comment by hynek** [comment]
OK I don't want to deal with this kind of crap. I'll finish this up as it is. We don't expose the ipaddress so it can be replaced by an inet_pton solution later if we run into actual problems.

**Comment by hynek** [comment]
Does this look good to you @twm?

File: tests/test_cryptography.py [hunk 1]
```diff
@@ -1,33 +1,73 @@
...
 
 
-class TestVerifyHostname(object):
-    def test_verify_hostname(self):
+class TestPublicAPI(object):
+    def test_certificate_verify_hostname_ok(self):
         """
-        It's just a convenience one-liner.  Let's check it doesn't explode b/c
-        of some typo.
+        verify_certificate_hostname succeeds if the hostname match.
```
Comment (twm) [inline tests/test_cryptography.py]: Typo: "match" &rarr; "matches"

**Comment by hynek** [comment]
here you go: https://pypi.org/project/service_identity/

**Pull request merged**